### PR TITLE
72_张子栋_第一次作业

### PR DESCRIPTION
**Pull request by Smillence**
add two functions to update employee and salary

### CODE REVIEW

**Comment by Smillence** [comment]
@annieliu

File: lesson1/assignment/payroll.sol [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        
+        require(msg.sender == employer);
+        
+        address currentEmployee = employee;
+
+        employee = newEmployee;
+        lastPayday = now;
+        
+        if (currentEmployee != 0x0) {
+            uint payment = salary * (now - lastPayday) / payDuration;
```
Comment (annieliu) [inline lesson1/assignment/payroll.sol]: Is this correct?

File: lesson1/assignment/payroll.sol [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    function updateSalary(uint newSalary) public {
+        require(msg.sender == employer);
+        
+        uint currentSalary = salary;
+        
+        salary = newSalary * 1 ether;
+        lastPayday = now;
+        
+        if (employee != 0x0) {
+            uint payment = currentSalary * (now - lastPayday) / payDuration;
```
Comment (annieliu) [inline lesson1/assignment/payroll.sol]: same here

File: lesson1/assignment/payroll.sol [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        
+        address currentEmployee = employee;
+        uint currentSalary = salary;
+        
+        employee = newEmployee;
+        salary = newSalary * 1 ether;
+        lastPayday = now;
+        
+        if (currentEmployee != 0x0) {
+            uint payment = currentSalary * (now - lastPayday) / payDuration;
```
Comment (annieliu) [inline lesson1/assignment/payroll.sol]: same here

File: lesson1/assignment/payroll.sol [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    function calculateRunway() public view returns (uint) {
+        return this.balance / salary;
+    }
+    
+    function hasEnoughFund() public view returns (bool) {
+        return calculateRunway() > 0;
+    }
+    
+    function getPaid() public {
+        uint nextPayday = lastPayday + payDuration;
```
Comment (annieliu) [inline lesson1/assignment/payroll.sol]: check it is the employee calling getPaid?

File: lesson1/assignment/payroll.sol (hunk 3 above)
Comment (Smillence) [reply lesson1/assignment/payroll.sol]: good catch! forgot to create currentLastPayday

File: lesson1/assignment/payroll.sol (hunk 4 above)
Comment (Smillence) [reply lesson1/assignment/payroll.sol]: yes I intentionally removed the require statement here to save some gas

btw, how much gas does `require(msg.sender == employee)` cost?

Comment (annieliu) [reply lesson1/assignment/payroll.sol]: 你可以去remix试验一下需要多少gas。讲道理从正确性的角度来说，应该加上。但是反正也是给employee发钱，我觉得不加也行。

**Pull request merged**